Incorrect ESM build

There are a few issues with the current ESM build output
See example of errors here https://codesandbox.io/p/devbox/confident-orla-xy9d9n?file=%2Findex.js

At runtime it seems it's not loading the .es.js version of the package
If I add
```json
// package.json
{
  "exports":{
    ".": {
      "import": "./dist/rand-seed.es.js",
      "require": "./dist/rand-seed.js",
      "types": "./dist/index.d.ts"
    }
  },
}
```
Then it correctly loads rand-seed.es.js. However since the package is commonjs, I then get other errors.
Would need to isolate rand-seed.es.js in another folder and add another package.json with just `{ "type": "module" }` in it.

Plus in typescript I get the following error
<img width="890" alt="image" src="https://github.com/user-attachments/assets/6c95f5e2-c5d5-4a8e-8266-75b3258a59ad">

I'm not sure how to fix all of this with rollup, so far I've only done dual commonjs/esm build with only typescript
Here's what I use for instance where the package is primarily ESM then I isolate the commonjs build

```json
// package.json
{
  "type": "module",
  "scripts": {
    "build": "yarn build:esm && yarn build:cjs",
    "build:esm": "yarn tsc --build",
    "build:cjs": "yarn tsc --build tsconfig.cjs.json && echo '{ \"type\": \"commonjs\" }' > dist-cjs/package.json"
  },
  "exports": {
    ".": {
      "types": {
        "import": "./dist/index.d.ts",
        "require": "./dist-cjs/index.d.ts"
      },
      "import": "./dist/index.js",
      "require": "./dist-cjs/index.js",
      "default": "./dist/index.js"
    }
  },
  "main": "./dist-cjs/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist",
    "dist-cjs",
    "package.json",
    "README.md",
    "CHANGELOG.md"
  ],
}
```

```json
// tsconfig.cjs.json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "tsBuildInfoFile": ".cache/.cjs.tsbuildinfo",
    "module": "commonjs",
    "moduleResolution": "node",
    "outDir": "dist-cjs",
  }
}
```

Yeah I updated my sandbox https://codesandbox.io/p/devbox/confident-orla-xy9d9n

I get 2 errors
1. `cjs/typescript.ts(1,28): error TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("rand-seed")' call instead.`
2. `node_modules/rand-seed/dist/index.d.ts(1,28): error TS2834: Relative import paths need explicit file extensions in ECMAScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.`

first issue is because typescript thinks the commonjs file `cjs/typescript.ts` is trying to import a purely ESM package (hence my dist-cjs solution)
second issue is the missing `.js` in `import Rand, { PRNG } from './Rand.js';`

I can't access your devbox (make sure it's public in the "share" section)

--- a/example/cjs/index.cjs
+++ b/example/cjs/index.cjs
@@ -1,10 +1,10 @@
-const Rand = require('../../dist/index.cjs').default;
-const PRNG = require('../../dist/index.cjs');
+const Rand = require('../../dist/cjs/index.js').default;
+const { PRNG } = require('../../dist/cjs/index.js');
 
 let rand = new Rand('1234', PRNG.mulberry32);
 const result1 = Array.from({ length: 10 }, () => rand.next());
 
-rand = new Rand('1234');
+rand = new Rand('1234', PRNG.mulberry32);
 const result2 = Array.from({ length: 10 }, () => rand.next());
 
 console.log('RESULT1:', result1);

--- a/example/es/index.js
+++ b/example/es/index.js
@@ -1,9 +1,9 @@
-import Rand, { PRNG } from '../../dist/index.js';
+import Rand, { PRNG } from '../../dist/es/index.js';
 
 let rand = new Rand('1234', PRNG.mulberry32);
 const result1 = Array.from({ length: 10 }, () => rand.next());
 
-rand = new Rand('1234');
+rand = new Rand('1234', PRNG.mulberry32);
 const result2 = Array.from({ length: 10 }, () => rand.next());
 
 console.log('RESULT1:', result1);

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "rand-seed",
-  "version": "2.1.0",
+  "version": "2.1.1",
   "description": "A small library for generating random numbers",
   "exports": {
     ".": {

--- a/package.json
+++ b/package.json
@@ -5,12 +5,12 @@
   "exports": {
     ".": {
       "import": {
-        "types": "./dist/index.d.ts",
-        "default": "./dist/index.js"
+        "types": "./dist/es/index.d.ts",
+        "default": "./dist/es/index.js"
       },
       "require": {
-        "types": "./dist/index.d.ts",
-        "default": "./dist/index.cjs"
+        "types": "./dist/cjs/index.d.ts",
+        "default": "./dist/cjs/index.js"
       }
     }
   },
@@ -18,7 +18,6 @@
   "files": [
     "dist"
   ],
-  "types": "dist/index.d.ts",
   "scripts": {
     "test": "cross-env NODE_ENV=test ./node_modules/.bin/jest",
     "lint": "npx eslint .",

--- a/rollup.config.ts
+++ b/rollup.config.ts
@@ -1,3 +1,5 @@
+import fs from 'fs';
+
 import typescript from '@rollup/plugin-typescript';
 import terser from '@rollup/plugin-terser';
 import clean from '@rollup-extras/plugin-clean';
@@ -10,14 +12,14 @@ export default defineConfig({
     output: [
         {
             dir: './dist',
-            entryFileNames: 'index.cjs',
+            entryFileNames: 'cjs/index.js',
             format: 'cjs',
             sourcemap: local,
             exports: 'named',
         },
         {
             dir: './dist',
-            entryFileNames: 'index.js',
+            entryFileNames: 'es/index.js',
             format: 'es',
             sourcemap: local,
         },
@@ -25,5 +27,49 @@ export default defineConfig({
     watch: {
         include: 'src/**',
     },
-    plugins: [clean('dist'), typescript({ sourceMap: local }), terser()],
+    plugins: [
+        clean('dist'),
+        typescript({ sourceMap: local }),
+        terser(),
+        (() => ({
+            name: 'collectCjsTypeDeclaration',
+            writeBundle: {
+                sequential: true,
+                order: 'post',
+                handler: ({ format }) => {
+                    if (format !== 'cjs') {
+                        return;
+                    }
+
+                    fs.readdir('./dist/es', { recursive: true }, (err, list) => {
+                        for (const item of list) {
+                            if (item.toString().endsWith('.js')) {
+                                continue;
+                            }
+
+                            if (fs.lstatSync(`./dist/es/${item.toString()}`).isDirectory()) {
+                                if (!fs.existsSync(`./dist/cjs/${item.toString()}`)) {
+                                    fs.mkdirSync(`./dist/cjs/${item.toString()}`, { recursive: true });
+                                }
+
+                                continue;
+                            }
+
+                            fs.copyFile(`./dist/es/${item}`, `./dist/cjs/${item}`, (err) => {
+                                if (err) {
+                                    console.error(err);
+                                }
+
+                                fs.writeFile('./dist/cjs/package.json', '{\n\t"type": "commonjs"\n}', (err) => {
+                                    if (err) {
+                                        console.error(err);
+                                    }
+                                });
+                            });
+                        }
+                    });
+                },
+            },
+        }))(),
+    ],
 });

--- a/src/Rand.ts
+++ b/src/Rand.ts
@@ -28,7 +28,7 @@ class Rand {
      * The string that will be used for generating a suitable hash for any of
      * the provided PRNG algorithms.
      *
-     * @var {string}
+     * @var {string|undefined}
      */
     private str?: string;
 
@@ -49,8 +49,8 @@ class Rand {
     /**
      * Create a new rand instance.
      *
-     * @param {string} str
-     * @param {PRNG} prng
+     * @param {string|undefined} str
+     * @param {PRNG|undefined} prng
      */
     public constructor(str?: string, prng: PRNG = PRNG.sfc32) {
         this.str = str;

--- a/tsconfig.base.json
+++ b/tsconfig.base.json
@@ -7,12 +7,11 @@
     "resolveJsonModule": true,
     "verbatimModuleSyntax": true,
     "useDefineForClassFields": true,
-    "esModuleInterop": true,
     "forceConsistentCasingInFileNames": true,
     "skipLibCheck": true,
     "strict": true,
     "declaration": true,
-    "declarationDir": "./dist",
+    "declarationDir": "./dist/es",
     "rootDir": "./src",
     "outDir": "./dist"
   },

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,3 +1,6 @@
 {
-  "extends": "./tsconfig.base"
+  "extends": "./tsconfig.base",
+  "compilerOptions": {
+    "esModuleInterop": false
+  }
 }

--- a/tsconfig.test.json
+++ b/tsconfig.test.json
@@ -1,6 +1,7 @@
 {
   "extends": "./tsconfig.base",
   "compilerOptions": {
+    "esModuleInterop": true,
     "verbatimModuleSyntax": false
   },
   "include": ["__tests__/**/*"]

